src/applications/oscilloscope.py

```diff
@@ -1,117 +1,247 @@
+        # https://stackoverflow.com/questions/53423868/matplotlib-animation-how-to-dynamically-extend-x-limits
+        # need to sporadically call
+        # fig.canvas.resize_event()
 
-        delta = 0.1*np.max(self.ydata)
+        delta = 0.1 * np.max(self.ydata)
         new_min = np.max([0, np.min(self.ydata) - delta])
         new_max = np.max(self.ydata) + delta
         current_min, current_max = self.ax.get_ylim()
-        if (np.abs((new_min - current_min)/(current_min)) > 0.12) or (np.abs((new_max - current_max)/(current_max)) > 0.12):
+
+        if (np.abs((new_min - current_min) / current_min) > 0.12) \
+                or (np.abs((new_max - current_max) / current_max) > 0.12):
             self.ax.set_ylim(np.max([0.01, np.min(self.ydata) - delta]), np.max(self.ydata) + delta)
+
         self.line.set_ydata(self.ydata)
-        return self.line,
 
+        if self.show_rolling_mean:
+            new_rm_value = self.update_rolling_mean()
+            self.rolling_mean_line.set_ydata(self.rolling_mean)
 
+        if self.show_reading_text:
+            text = self._get_text_value(y)
+            if self.show_rolling_mean:
+                text = f'{text}\n {self._get_text_value(new_rm_value)}'
+                measured_stdev = self.get_new_rolling_stdev_val()
+                expected_stdev = np.sqrt(new_rm_value)
+                text = f'{text}\n {self._get_text_value(measured_stdev)} ({self._get_text_value(expected_stdev)})'
+            self.current_value_text.set_text(text)
 
-class MainApplicationView():
-    def __init__(self, main_frame):
+        return self.line,
+
+    def get_new_rolling_mean_val(self):
+        """ Calculates new rolling mean value. """
+        values_of_interest = list(self.ydata)[len(self.ydata) - 2 * self.half_window_size:
+                                              len(self.ydata)]
+        return np.mean(values_of_interest)
+
+    def get_new_rolling_stdev_val(self):
+        """ Calculates new rolling standard deviation. """
+        values_of_interest = list(self.ydata)[len(self.ydata) - 2 * self.half_window_size:
+                                              len(self.ydata)]
+        return np.std(values_of_interest)
+
+    def update_rolling_mean(self):
+        """ Updates internal rolling mean buffer. """
+        rm = collections.deque(self.rolling_mean[self.half_window_size: len(self.ydata) - self.half_window_size])
+        rm.popleft()
+        new_rm_value = self.get_new_rolling_mean_val()
+        rm.append(new_rm_value)
+        self.rolling_mean[self.half_window_size: len(self.ydata) - self.half_window_size] = list(rm)
+        return new_rm_value
+
+    def get_rolling_mean(self):
+        """ Computes rolling mean of current data. """
+        rolling_mean = get_rolling_mean(self.ydata, self.rolling_mean_window_size)
+        rolling_mean[:self.half_window_size] = np.nan
+        rolling_mean[len(self.ydata) - self.half_window_size:] = np.nan
+        return rolling_mean
+
```

Comment:
See message above asking if we need to store a self.rolling_mean list in memory and perform this popleft / append step every time. It makes sense if using scipy to compute the full rolling mean is slow.